File: C++Lib/hsgmParser.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <stdexcept>

#include <png.h>

class Binding {
public:
    std::unordered_map<std::string, std::string> textureBindings;
    std::unordered_map<std::string, std::string> colliderBindings;
    std::unordered_map<std::string, std::string> triggerBindings;

    std::string GetTexturePath(const std::string& textureName) const {
        auto it = textureBindings.find(textureName);
        return it != textureBindings.end() ? it->second : "";
    }

    std::string GetColliderColor(const std::string& colliderType) const {
        auto it = colliderBindings.find(colliderType);
        return it != colliderBindings.end() ? it->second : "";
    }

    std::string GetTriggerColor(const std::string& triggerType) const {
        auto it = triggerBindings.find(triggerType);
        return it != triggerBindings.end() ? it->second : "";
    }
};

void ParseMapFile(const std::string& filePath, Binding& binding,
                  std::unordered_map<std::string, std::pair<int, int>>& textures,
                  std::unordered_map<std::string, std::pair<int, int>>& colliders,
                  std::unordered_map<std::string, std::pair<int, int>>& triggers) {
    std::ifstream file(filePath);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open file.");
    }

    std::unordered_map<std::string, std::string> textureBindings;
    std::unordered_map<std::string, std::string> colliderBindings;
    std::unordered_map<std::string, std::string> triggerBindings;

    std::string line;
    while (std::getline(file, line)) {
        size_t equalPos = line.find('=');
        if (equalPos == std::string::npos) continue;

        std::string key = line.substr(0, equalPos);
        std::string value = line.substr(equalPos + 1);
        value.erase(0, value.find_first_not_of(" \t"));
        value.erase(value.find_last_not_of(" \t") + 1);

        if (key.find("Texture") != std::string::npos) {
            textures[key] = ParseTuple(value);
        } else if (key.find("Collider") != std::string::npos) {
            colliders[key] = ParseTuple(value);
        } else if (key.find("Trigger") != std::string::npos) {
            triggers[key] = ParseTuple(value);
        } else if (key.find("Binding") != std::string::npos) {
            auto bindingParts = Split(key, ':');
            if (bindingParts.size() == 2) {
                std::string type = bindingParts[0];
                std::string name = bindingParts[1];
                if (type == "Texture") {
                    textureBindings[name] = value;
                } else if (type == "Collider") {
                    colliderBindings[name] = value;
                } else if (type == "Trigger") {
                    triggerBindings[name] = value;
                }
            }
        }
    }

    binding.textureBindings = textureBindings;
    binding.colliderBindings = colliderBindings;
    binding.triggerBindings = triggerBindings;
}

std::pair<int, int> ParseTuple(const std::string& str) {
    size_t commaPos = str.find(',');
    if (commaPos == std::string::npos) throw std::runtime_error("Invalid tuple format.");

    int x = std::stoi(str.substr(0, commaPos));
    int y = std::stoi(str.substr(commaPos + 1));
    return { x, y };
}

void WritePng(const std::string& filePath, int width, int height, const std::vector<unsigned char>& pixelData) {
    FILE *fp = fopen(filePath.c_str(), "wb");
    if (!fp) throw std::runtime_error("Failed to open file.");

    png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
    if (!png) {
        fclose(fp);
        throw std::runtime_error("Failed to create PNG write structure.");
    }

    png_infop info = png_create_info_struct(png);
    if (!info) {
        png_destroy_write_struct(&png, nullptr);
        fclose(fp);
        throw std::runtime_error("Failed to create PNG info structure.");
    }

    if (setjmp(png_jmpbuf(png))) {
        png_destroy_write_struct(&png, &info);
        fclose(fp);
        throw std::runtime_error("Error during PNG creation.");
    }

    png_init_io(png, fp);
    png_set_IHDR(
        png,
        info,
        width, height,
        8, PNG_COLOR_TYPE_RGBA, PNG_INTERLACE_NONE,
        PNG_COMPRESSION_TYPE_DEFAULT,
        PNG_FILTER_TYPE_DEFAULT
    );

    png_write_info(png, info);

    std::vector<png_bytep> rows(height);
    for (int y = 0; y < height; ++y) {
        rows[y] = const_cast<png_bytep>(&pixelData[y * width * 4]);
    }
    png_write_image(png, rows.data());
    png_write_end(png, nullptr);

    png_destroy_write_struct(&png, &info);
    fclose(fp);
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: MapParserAndPngEncoder <mapFile> <outputPng>" << std::endl;
        return 1;
    }

    std::string mapFile = argv[1];
    std::string outputPng = argv[2];

    try {
        Binding binding;
        std::unordered_map<std::string, std::pair<int, int>> textures;
        std::unordered_map<std::string, std::pair<int, int>> colliders;
        std::unordered_map<std::string, std::pair<int, int>> triggers;

        // Parse the map file
        ParseMapFile(mapFile, binding, textures, colliders, triggers);

        // Generate a basic image data array
        int width = 256;
        int height = 256;
        std::vector<unsigned char> pixelData(width * height * 4); // RGBA

        // Fill image data with a simple pattern for demonstration
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                int index = (y * width + x) * 4;
                pixelData[index] = static_cast<unsigned char>(x % 256);       // Red
                pixelData[index + 1] = static_cast<unsigned char>(y % 256);   // Green
                pixelData[index + 2] = 0;                                      // Blue
                pixelData[index + 3] = 255;                                    // Alpha
            }
        }

        // Save as PNG
        WritePng(outputPng, width, height, pixelData);

        std::cout << "PNG file generated: " << outputPng << std::endl;
    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << std::endl;
        return 1;
    }

    return 0;
}

File: C#Lib/hsgmParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace MapUtilities
{
    public static class MapParserAndPngEncoder
    {
        private const int TileSize = 16;
        private static readonly byte[] PngSignature = { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A };

        public class Binding
        {
            private readonly Dictionary<string, string> textureBindings;
            private readonly Dictionary<string, string> colliderBindings;
            private readonly Dictionary<string, string> triggerBindings;

            public Binding(Dictionary<string, string> textureBindings,
                           Dictionary<string, string> colliderBindings,
                           Dictionary<string, string> triggerBindings)
            {
                this.textureBindings = textureBindings;
                this.colliderBindings = colliderBindings;
                this.triggerBindings = triggerBindings;
            }

            public string GetTexturePath(string textureName)
            {
                return textureBindings.TryGetValue(textureName, out var path) ? path : null;
            }

            public string GetColliderColor(string colliderType)
            {
                return colliderBindings.TryGetValue(colliderType, out var color) ? color : null;
            }

            public string GetTriggerColor(string triggerType)
            {
                return triggerBindings.TryGetValue(triggerType, out var color) ? color : null;
            }
        }

        public static void ParseMapFile(string filePath, out Binding binding,
                                        out Dictionary<string, Tuple<int, int>> textures,
                                        out Dictionary<string, Tuple<int, int>> colliders,
                                        out Dictionary<string, Tuple<int, int>> triggers)
        {
            textures = new Dictionary<string, Tuple<int, int>>();
            colliders = new Dictionary<string, Tuple<int, int>>();
            triggers = new Dictionary<string, Tuple<int, int>>();

            var textureBindings = new Dictionary<string, string>();
            var colliderBindings = new Dictionary<string, string>();
            var triggerBindings = new Dictionary<string, string>();

            foreach (var line in File.ReadLines(filePath))
            {
                var parts = line.Split('=');
                if (parts.Length != 2) continue;

                var key = parts[0].Trim();
                var value = parts[1].Trim().Trim('[', ']').Split(',');

                if (key.StartsWith("Texture"))
                    textures[key] = ParseTuple(value);
                else if (key.StartsWith("Collider"))
                    colliders[key] = ParseTuple(value);
                else if (key.StartsWith("Trigger"))
                    triggers[key] = ParseTuple(value);
                else if (key.StartsWith("Binding"))
                {
                    var bindingParts = key.Split(':');
                    if (bindingParts.Length == 2)
                    {
                        var type = bindingParts[0].Trim();
                        var name = bindingParts[1].Trim();
                        if (type == "Texture")
                            textureBindings[name] = value[0].Trim();
                        else if (type == "Collider")
                            colliderBindings[name] = value[0].Trim();
                        else if (type == "Trigger")
                            triggerBindings[name] = value[0].Trim();
                    }
                }
            }

            binding = new Binding(textureBindings, colliderBindings, triggerBindings);
        }

        private static Tuple<int, int> ParseTuple(string[] values)
        {
            if (values.Length != 2) throw new FormatException("Invalid tuple format.");
            return new Tuple<int, int>(int.Parse(values[0]), int.Parse(values[1]));
        }

        public static void SaveAsPng(int width, int height, byte[] pixelData, string filePath)
        {
            using (var stream = new FileStream(filePath, FileMode.Create))
            using (var writer = new BinaryWriter(stream))
            {
                WritePngSignature(writer);
                WriteIHDRChunk(writer, width, height);
                WriteIDATChunk(writer, pixelData);
                WriteIENDChunk(writer);
            }
        }

        private static void WritePngSignature(BinaryWriter writer)
        {
            writer.Write(PngSignature);
        }

        private static void WriteIHDRChunk(BinaryWriter writer, int width, int height)
        {
            writer.Write(13); // Length of IHDR chunk data
            writer.Write(Encoding.ASCII.GetBytes("IHDR"));
            writer.Write(width);
            writer.Write(height);
            writer.Write((byte)8); // Bit depth
            writer.Write((byte)6); // Color type: RGBA
            writer.Write((byte)0); // Compression method
            writer.Write((byte)0); // Filter method
            writer.Write((byte)0); // Interlace method
            WriteChunkCRC(writer, "IHDR");
        }

        private static void WriteIDATChunk(BinaryWriter writer, byte[] pixelData)
        {
            using (var idatStream = new MemoryStream())
            using (var idatWriter = new BinaryWriter(idatStream))
            {
                WriteImageData(idatWriter, pixelData);
                idatStream.Position = 0;

                // Compress the IDAT chunk data using DEFLATE
                using (var compressedStream = new MemoryStream())
                using (var deflateStream = new System.IO.Compression.DeflateStream(compressedStream, System.IO.Compression.CompressionMode.Compress, true))
                {
                    idatStream.CopyTo(deflateStream);
                }

                var idatData = compressedStream.ToArray();
                writer.Write(idatData.Length); // Length of IDAT chunk data
                writer.Write(Encoding.ASCII.GetBytes("IDAT"));
                writer.Write(idatData);
                WriteChunkCRC(writer, "IDAT");
            }
        }

        private static void WriteImageData(BinaryWriter writer, byte[] pixelData)
        {
            writer.Write((byte)0); // No filter
            writer.Write(pixelData);
        }

        private static void WriteIENDChunk(BinaryWriter writer)
        {
            writer.Write(0); // Length of IEND chunk data
            writer.Write(Encoding.ASCII.GetBytes("IEND"));
            WriteChunkCRC(writer, "IEND");
        }

        private static void WriteChunkCRC(BinaryWriter writer, string chunkType)
        {
            uint crc = Crc32(Encoding.ASCII.GetBytes(chunkType));
            writer.Write(crc);
        }

        private static uint Crc32(byte[] data)
        {
            const uint Polynomial = 0xedb88320;
            uint[] table = new uint[256];
            for (uint i = 0; i < 256; ++i)
            {
                uint crc = i;
                for (uint j = 8; j > 0; --j)
                {
                    if ((crc & 1) == 1)
                    {
                        crc = (crc >> 1) ^ Polynomial;
                    }
                    else
                    {
                        crc >>= 1;
                    }
                }
                table[i] = crc;
            }

            uint crcValue = 0xffffffff;
            foreach (byte b in data)
            {
                byte tableIndex = (byte)((crcValue & 0xff) ^ b);
                crcValue = (crcValue >> 8) ^ table[tableIndex];
            }

            return crcValue ^ 0xffffffff;
        }

        public static void Main(string[] args)
        {
            if (args.Length != 2)
            {
                Console.WriteLine("Usage: MapParserAndPngEncoder <mapFile> <outputPng>");
                return;
            }

            string mapFile = args[0];
            string outputPng = args[1];

            try
            {
                // Parse the map file
                ParseMapFile(mapFile, out var binding, out var textures, out var colliders, out var triggers);

                // Generate a basic image data array
                int width = 256;
                int height = 256;
                byte[] pixelData = new byte[width * height * 4]; // RGBA

                // Fill image data with a simple pattern for demonstration
                for (int y = 0; y < height; y++)
                {
                    for (int x = 0; x < width; x++)
                    {
                        int index = (y * width + x) * 4;
                        pixelData[index] = (byte)(x % 256);       // Red
                        pixelData[index + 1] = (byte)(y % 256);   // Green
                        pixelData[index + 2] = 0;                  // Blue
                        pixelData[index + 3] = 255;                // Alpha
                    }
                }

                // Save as PNG
                SaveAsPng(width, height, pixelData, outputPng);

                Console.WriteLine($"PNG file generated: {outputPng}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
            }
        }
    }
}
